synchronization/trigger_test.go

```diff
@@ -26,120 +26,133 @@ func newNopLogger() *nopLogger {
-
+		for i := 0; i < 5; i++ {
+			ch <- time.Now()
+			triggerFlag.EventuallyToggled(fmt.Sprintf("expected %d trigger invocations", i))
+		}
 	})
 }
 
-func TestPeriodicalTriggerStopOnContextCancel(t *testing.T) {
+func TestPeriodicalTriggerKeepsTriggeringAfterPanic(t *testing.T) {
 	with.Concurrency(t, func(parent context.Context, harness *with.ConcurrencyHarness) {
-		ctx, cancel := context.WithCancel(parent)
-		x := 0
-		p := synchronization.NewPeriodicalTrigger(ctx, "a periodical trigger", time.Millisecond*2, harness.Logger, func() { x++ }, nil)
+		logger := newNopLogger()
+		ticker := synchronization.NewHookTicker()
+		ch := make(chan time.Time)
+		ticker.C_ = func() <-chan time.Time { return ch }
+		triggerFlag := NewTestFlag(t)
+		p := synchronization.NewPeriodicalTrigger(context.Background(), "a periodical trigger", ticker, logger, func() {
+			triggerFlag.Toggle()
+			panic("we should not see this other than the logs")
+		}, nil)
+		defer p.Stop()
 		harness.Supervise(p)
-		cancel()
-		time.Sleep(3 * time.Millisecond)
-		require.Equal(t, 0, x, "expected no ticks")
+		for i := 0; i < 5; i++ {
+			ch <- time.Now()
+			triggerFlag.EventuallyToggled(fmt.Sprintf("expected %d trigger invocations even though it panics", i))
+		}
 	})
 }
 
-func TestPeriodicalTriggerStopWorksWhenContextIsCancelled(t *testing.T) {
-	with.Concurrency(t, func(parent context.Context, harness *with.ConcurrencyHarness) {
-		ctx, cancel := context.WithCancel(parent)
-		cancel() // send a cancelled context to reduce chances of trigger being called even once
-		x := 0
-		p := synchronization.NewPeriodicalTrigger(ctx, "a periodical trigger", time.Millisecond*2, harness.Logger, func() { x++ }, nil)
+func TestPeriodicalTriggerStopsTickerOnCallToStop(t *testing.T) {
+	with.Concurrency(t, func(ctx context.Context, harness *with.ConcurrencyHarness) {
+		ticker := synchronization.NewHookTicker()
+		wasStopped := NewTestFlag(t)
+		ticker.Stop_ = func() { wasStopped.Toggle() }
+		p := synchronization.NewPeriodicalTrigger(context.Background(), "a periodical trigger", ticker, harness.Logger, func() {}, nil)
 		harness.Supervise(p)
-		time.Sleep(3 * time.Millisecond)
-		require.Equal(t, 0, x, "expected no ticks")
 		p.Stop()
-		require.Equal(t, 0, x, "expected stop to not block")
-
+		wasStopped.EventuallyToggled("expected ticker.Stop() to have been called")
 	})
 }
 
-func TestPeriodicalTriggerStopOnContextCancelWithStopAction(t *testing.T) {
+func TestPeriodicalTriggerStopsTickerOnContextCancel(t *testing.T) {
 	with.Concurrency(t, func(parent context.Context, harness *with.ConcurrencyHarness) {
 		ctx, cancel := context.WithCancel(parent)
-		ch := make(chan struct{})
-		p := synchronization.NewPeriodicalTrigger(ctx, "a periodical trigger", time.Millisecond*2, harness.Logger, func() {}, func() { close(ch) })
+		ticker := synchronization.NewHookTicker()
+		wasStopped := NewTestFlag(t)
+		ticker.Stop_ = func() { wasStopped.Toggle() }
```

Comment:
it seems like you use `Stop_` *or* `C_` but never both. If so, I'd either create separate make functions `NewHookTriggerWithChannel()` and `NewHookTriggerWithStopFunc(f)` and use either when appropriate.